File: python/reporting.py
class FlowchartReportItems:
    def __init__(self, name: str, serial_number: str, filename: str):
        # this implementation is deliberately blank - can't use this class in a unit test
        pass

    def add(self, summary_information: str) -> None:
        raise NotImplementedError("can't call from a unit test")

    def to_array(self) -> list:
        raise NotImplementedError("can't call from a unit test")


class DiagramPage:
    def __init__(self, name: str, page_contents: str):
        self._name = name
        self._page_contents = page_contents

    @property
    def name(self) -> str:
        return self._name

    @property
    def page_contents(self) -> str:
        return self._page_contents

    def add_data_from(self, data: FlowchartReportItems) -> None:
        try:
            if "{0}" in self._page_contents:
                self._page_contents = self._page_contents.format(*data.to_array())
        except ValueError:
            # Silently handle formatting errors
            pass


class PagesBuilder:
    def process_page(self, page: DiagramPage, data: FlowchartReportItems) -> 'DiagramReportPage':
        page.add_data_from(data)
        return DiagramReportPage(page)

    def apply(self, report: 'DiagramPagesReport', report_pages: list['DiagramReportPage']) -> bool:
        if report.is_complete():
            return False

        for page in report_pages:
            report.add_page(page)

        return True


class DiagramReportPage:
    def __init__(self, page):
        self.pages = [page]


class DiagramPagesReport:
    def __init__(self):
        self._pages: list['DiagramReportPage'] = []

    def add_page(self, page: 'DiagramReportPage') -> None:
        self._pages.append(page)

    def is_complete(self) -> bool:
        return len(self._pages) > 0

File: typescript/src/reporting.ts
export class FlowchartReportItems {
  constructor(name: string, serialNumber: string, filename: string) {
    // this implementation is deliberately blank - can't use this class in a unit test
  }

  add(summaryInformation: string): void {
    throw new Error("can't call from a unit test");
  }

  toArray(): Array<any> {
    throw new Error("can't call from a unit test");
  }
}

export class DiagramPage {
  private readonly _name: string;
  private _pageContents: string;

  constructor(name: string, pageContents: string) {
    this._name = name;
    this._pageContents = pageContents;
  }

  get name(): string {
    return this._name;
  }

  get pageContents(): string {
    return this._pageContents;
  }

  addDataFrom(data: FlowchartReportItems): void {
    try {
      if (this._pageContents.includes("{0}")) {
        this._pageContents = this._pageContents.replace(/\{(\d+)\}/g, (...args) => {
          const index = parseInt(args[1]);
          const dataArray = data.toArray();
          return dataArray[index]?.toString() ?? '';
        });
      }
    } catch (error) {
      // Silently handle formatting errors
    }
  }
}

export class PagesBuilder {
  processPage(page: DiagramPage, data: FlowchartReportItems): DiagramReportPage {
    page.addDataFrom(data);
    return new DiagramReportPage(page);
  }

  apply(report: DiagramPagesReport, reportPages: DiagramReportPage[]): boolean {
    if (report.isComplete()) {
      return false;
    }

    for (const page of reportPages) {
      report.addPage(page);
    }

    return true;
  }
}

export class DiagramReportPage {
  pages: DiagramPage[];

  constructor(page: DiagramPage) {
    this.pages = [page];
  }
}

export class DiagramPagesReport {
  private _pages: DiagramReportPage[] = [];

  addPage(page: DiagramReportPage): void {
    this._pages.push(page);
  }

  isComplete(): boolean {
    return this._pages.length > 0;
  }
}

export class FlowchartReport {
  close(): void {
    throw new Error("Not supported");
  }

  open(show: boolean): void {
    throw new Error("Not supported");
  }

  isOpen(): boolean {
    throw new Error("Not supported");
  }

  openWithContents(reportTemplate: string, substitutions: FlowchartReportItems, readOnly: boolean): void {
    throw new Error("Not supported");
  }

  saveToFile(targetFilename: string): void {
    throw new Error("Not supported");
  }
}
